static/js/tools/timeline_page.tsx

```diff
@@ -17,119 +17,122 @@
+  statsVarPaths: string[][];
   statsVarInfo: { [key: string]: StatsVarInfo };
-  places: [string, string][]; // [(placeId, placeName)]
+  placeIdNames: Record<string, string>; // [(placeId, placeName)]
+  statsVarTitle: Record<string, string>;
   perCapita: boolean;
   statsVarValid: Set<string>;
 }
 
-class Page extends Component<PagePropType, PageStateType> {
-  constructor(props: PagePropType) {
+class Page extends Component<Record<string, unknown>, PageStateType> {
+  params: TimelineParams;
+
+  constructor(props: Record<string, unknown>) {
     super(props);
-    this.handleHashChange = this.handleHashChange.bind(this);
-    this._togglePerCapita = this._togglePerCapita.bind(this);
-    this.setStatsVarTitle = this.setStatsVarTitle.bind(this);
+    this.params = getTimelineParamsFromUrl();
+    const tempStatsVarTitle = {};
+    for (const statsVar of this.params.getStatsVarDcids()) {
+      tempStatsVarTitle[statsVar] = statsVar;
+    }
     this.state = {
-      statsVarPaths: [],
-      statsVarInfo: {},
-      places: [],
-      perCapita: false,
+      placeIdNames: {},
       statsVarValid: new Set(),
+      statsVarPaths: this.params.getStatsVarPaths(),
+      statsVarInfo: {},
+      statsVarTitle: tempStatsVarTitle,
+      perCapita: this.params.pc,
     };
   }
-  shouldComponentUpdate(
-    nextProps: PagePropType,
-    nextState: PageStateType
-  ): boolean {
-    return (
-      JSON.stringify(this.state.statsVarInfo) !==
-        JSON.stringify(nextState.statsVarInfo) ||
-      this.state.places !== nextState.places ||
-      this.state.perCapita !== nextState.perCapita
-    );
-  }
 
   componentDidMount(): void {
-    window.addEventListener("hashchange", this.handleHashChange);
-    this.handleHashChange();
-  }
-
-  handleHashChange(): void {
-    const urlVar = parseUrl();
-    let statsVarInfoPromise = Promise.resolve(this.state.statsVarInfo);
-    if (urlVar.statsVarPath !== this.state.statsVarPaths) {
-      if (urlVar.statsVarId.length !== 0) {
-        statsVarInfoPromise = getStatsVarInfo(urlVar.statsVarId);
-      } else {
-        statsVarInfoPromise = Promise.resolve({});
-      }
-    }
-
-    let placesPromise = Promise.resolve(this.state.places);
-    let validStatsVarPromise = Promise.resolve(this.state.statsVarValid);
-    if (urlVar.placeId !== Object.keys(this.state.places)) {
-      validStatsVarPromise = getStatsVar(urlVar.placeId);
-      if (urlVar.placeId.length !== 0) {
-        placesPromise = getPlaceNames(urlVar.placeId).then((data) =>
-          Object.entries(data)
-        );
-      } else {
-        placesPromise = Promise.resolve([]);
-      }
-    }
-
+    const statsVarInfoPromise = getStatsVarInfo(this.params.getStatsVarDcids());
+    const placesPromise = getPlaceNames(this.params.placeDcids);
+    const validStatsVarPromise = getStatsVar(this.params.placeDcids);
     Promise.all([
       statsVarInfoPromise,
       placesPromise,
       validStatsVarPromise,
     ]).then((values) => {
```

Comment:
nit: here "values" can be written as [statsVarInfo, placeIdNames, statsVarValid]

Reply:
done.